pkg/controller/daemonpodupdater/daemon_pod_updater_controller.go

```diff
@@ -0,0 +1,507 @@
+	defer c.daemonsetWorkqueue.ShutDown()
+
+	// Synchronize the cache before starting to process events
+	if !cache.WaitForCacheSync(stopCh, c.daemonsetSynced, c.nodeSynced,
+		c.podSynced) {
+		klog.Error("sync daemonPodUpdater controller timeout")
+	}
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (c *Controller) runWorker() {
+	for {
+		obj, shutdown := c.daemonsetWorkqueue.Get()
+		if shutdown {
+			return
+		}
+
+		// TODO(hxc): should requeue failed ds
+		if err := c.syncHandler(obj.(string)); err != nil {
+			utilruntime.HandleError(err)
+		}
+		c.daemonsetWorkqueue.Done(obj)
+	}
+}
+
+func (c *Controller) syncHandler(key string) error {
+	defer func() {
+		klog.V(4).Infof("Finish syncing daemonPodUpdater request %q", key)
+	}()
+
+	klog.V(4).Infof("Start handling daemonPodUpdater request %q", key)
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return fmt.Errorf("invalid resource key: %s", key)
+	}
+
+	// Daemonset that need to be synced
+	ds, err := c.daemonsetLister.DaemonSets(namespace).Get(name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			c.expectations.DeleteExpectations(key)
+			return nil
+		}
+		return err
+	}
+
+	if ds.DeletionTimestamp != nil {
+		return nil
+	}
+
+	// Only process daemonset that meets expectations
+	// Otherwise, wait native daemonset controller reconciling
+	if !c.expectations.SatisfiedExpectations(key) {
+		return nil
+	}
+
+	nodeToDaemonPods, err := c.getNodesToDaemonPods(ds)
+	if err != nil {
+		return fmt.Errorf("couldn't get node to daemon pod mapping for daemon set %q: %v", ds.Name, err)
+	}
+
+	// Recheck required annotation
+	v, ok := ds.Annotations[UpdateAnnotation]
+	if !ok {
+		return fmt.Errorf("won't sync daemonset %q without annotation 'apps.openyurt.io/update-strategy'", ds.Name)
+	}
+
+	switch v {
+	case OTAUpdate:
+		if err := c.checkOTAUpdate(ds, nodeToDaemonPods); err != nil {
+			return err
+		}
+
+	case AutoUpdate:
+		if err := c.autoUpdate(ds, nodeToDaemonPods); err != nil {
+			return err
+		}
+	default:
+		return fmt.Errorf("unknown annotation type %v", v)
+	}
+
+	return nil
+}
+
+// checkOTAUpdate compare every pod to its owner daemonset to check if pod is updatable
+// If pod is in line with the latest daemonset spec, set annotation "apps.openyurt.io/pod-updatable" to "true"
+// while not, set annotation "apps.openyurt.io/pod-updatable" to "false"
+func (c *Controller) checkOTAUpdate(ds *appsv1.DaemonSet, nodeToDaemonPods map[string][]*corev1.Pod) error {
+	for nodeName, pods := range nodeToDaemonPods {
+		// Check if node is ready, ignore not-ready node
+		ready, err := NodeReadyByName(c.nodeLister, nodeName)
+		if err != nil {
+			return fmt.Errorf("couldn't check node %q ready status, %v", nodeName, err)
+		}
+		if !ready {
+			continue
+		}
+		for _, pod := range pods {
+			if err := SetPodUpdateAnnotation(c.kubeclientset, ds, pod); err != nil {
```

Comment:
it's not a good idea to a new annotation to specify that pod should be upgraded. I think we can add a new pod status condition for pod to specify pod upgrade, and the info as following:
```
condition type: PodNeedUpdate
condition status: True/False
```

Reply:
`PodNeedUpgrade` or `PodNeedUpdate` ?
I found I got `PodNeedUpgrade` in github comment while `PodNeedUpdate` in email notification.

Reply:
> `PodNeedUpgrade` or `PodNeedUpdate` ? I found I got `PodNeedUpgrade` in github comment while `PodNeedUpdate` in email notification.

I have modified `PodNeedUpdate` to `PodNeedUpgrade`. so please use `PodNeedUpgrade`.